src/main/java/com/minecolonies/coremod/entity/ai/citizen/hundertaker/EntityAIWorkUndertaker.java

```diff
@@ -0,0 +1,517 @@
+
+    /**
+     * Undertaker digging icon
+     */
+    private final static VisibleCitizenStatus DIGGING_ICON =
+            new VisibleCitizenStatus(new ResourceLocation(Constants.MOD_ID, "textures/icons/work/undertaker.png"), "com.minecolonies.gui.visiblestatus.digging");
+
+    /**
+     * Undertaker bury icon
+     */
+    private final static VisibleCitizenStatus BURYING_ICON =
+            new VisibleCitizenStatus(new ResourceLocation(Constants.MOD_ID, "textures/icons/work/undertaker.png"), "com.minecolonies.gui.visiblestatus.burying");
+
+    /**
+     * Undertaker resurrect icon
+     */
+    private final static VisibleCitizenStatus RESURRECT_ICON =
+            new VisibleCitizenStatus(new ResourceLocation(Constants.MOD_ID, "textures/icons/work/undertaker.png"), "com.minecolonies.gui.visiblestatus.resurrect");
+
+    /**
+     * Changed after finished digging in order to dump the inventory.
+     */
+    private boolean shouldDumpInventory = false;
+
+    /**
+     * The current pos to wander at.
+     */
+    private BlockPos wanderPos = null;
+
+    /**
+     * The current pos to grave to build.
+     */
+    private Tuple<BlockPos, Direction> burialPos = null;
+
+    /**
+     * Constructor for the Undertaker. Defines the tasks the Undertaker executes.
+     *
+     * @param job a undertaker job to use.
+     */
+    public EntityAIWorkUndertaker(@NotNull final JobUndertaker job)
+    {
+        super(job);
+        super.registerTargets(
+          new AITarget(IDLE, START_WORKING, REQUEST_DELAY),
+          new AITarget(START_WORKING, this::startWorking, STANDARD_DELAY),
+          new AITarget(WANDER, this::wander, STANDARD_DELAY),
+          new AITarget(EMPTY_GRAVE, this::emptyGrave, STANDARD_DELAY),
+          new AITarget(TRY_RESURRECT, this::tryResurrect, STANDARD_DELAY),
+          new AITarget(DIG_GRAVE, this::digGrave, STANDARD_DELAY),
+          new AITarget(BURY_CITIZEN, this::buryCitizen, STANDARD_DELAY)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    @Override
+    public Class<BuildingGraveyard> getExpectedBuildingClass()
+    {
+        return BuildingGraveyard.class;
+    }
+
+    /**
+     * Prepares the undertaker for digging. Also requests the tools and checks if the undertaker has queued graves.
+     *
+     * @return the next IAIState
+     */
+    @NotNull
+    private IAIState startWorking()
+    {
+        @Nullable final BuildingGraveyard building = getOwnBuilding();
+        if (building == null || building.getBuildingLevel() < 1)
+        {
+            worker.getCitizenData().setVisibleStatus(null);
+            return IDLE;
+        }
+
+        worker.getCitizenData().setVisibleStatus(VisibleCitizenStatus.WORKING);
+        worker.getCitizenData().setIdleAtJob(false);
+
+        @Nullable final BlockPos currentGrave = building.getGraveToWorkOn();
+        if (currentGrave != null)
+        {
+            if (walkToBuilding())
```

Comment:
shouldn't this be walkToPos ?

Reply:
In this case, the undertaker first go back to the graveyard to take the order as he may be wandering around in the city. This is just cosmetic but feels more natural to me when I playtested. If a grave appear, the Undertaker first have to go back to the graveyard to "get ready to dig" then goes to the grave

Reply:
okay, then we need some small changes
a) first walk to building, then getGraveToWorkOn should check if the chunk of the tileEntity is loaded

Reply:
(We should always only check for a block or tileentity when we stand in front of it, to avoid chunk loading)